thetis/shallowwater_eq.py

```diff
@@ -998,3 +1001,455 @@ def residual(self, label, solution, solution_old, fields, fields_old, bnd_condit
+        self.add_term(AtmosphericPressureResidual(*args), 'source')
+        self.add_term(QuadraticDragResidual(*args), 'explicit')
+        self.add_term(LinearDragResidual(*args), 'explicit')
+        self.add_term(BottomDrag3DResidual(*args), 'source')
+        self.add_term(TurbineDragResidual(*args), 'implicit')
+        self.add_term(MomentumSourceResidual(*args), 'source')
+
+    def add_continuity_terms(self, *args):
+        self.add_term(HUDivResidual(*args), 'implicit')
+        self.add_term(ContinuitySourceResidual(*args), 'source')
+
+    def residual_cell(self, label, uv, eta, uv_old, eta_old, fields, fields_old, bnd_conditions):
+        f = 0
+        print("\nCell residual norm contributions:")
+        for term in self.select_terms(label):
+            r = term.residual_cell(uv, eta, uv_old, eta_old, fields, fields_old, bnd_conditions)
+            if r is not None:
+                f += r
+                print("    {name:30s} {norm:.4e}".format(name=term.name, norm=norm(r)))
+        return f
+
+    def residual_edge(self, label, uv, eta, uv_old, eta_old, fields, fields_old, bnd_conditions):
+        f = 0
+        print("\nEdge residual norm contributions:")
+        for term in self.select_terms(label):
+            r = term.residual_edge(uv, eta, uv_old, eta_old, fields, fields_old, bnd_conditions)
+            if r is not None:
+                f += r
+                print("    {name:30s} {norm:.4e}".format(name=term.name, norm=norm(r)))
+        return f
+
+
+class FreeSurfaceResidual(BaseShallowWaterResidual):
+    """
+    2D free surface equation :eq:`swe_freesurf` in non-conservative form.
+    """
+    def __init__(self, eta_space, u_space,
+                 bathymetry,
+                 options):
+        """
+        :arg eta_test: test function of the elevation function space
+        :arg u_space: velocity function space
+        :arg function_space: Mixed function space where the solution belongs
+        :arg bathymetry: bathymetry of the domain
+        :type bathymetry: :class:`Function` or :class:`Constant`
+        :arg options: :class:`.AttrDict` object containing all circulation model options
+        """
+        super(FreeSurfaceResidual, self).__init__(eta_space, bathymetry, options)
+        self.add_continuity_terms(eta_space, u_space, bathymetry, options)
+
+    def cell_residual(self, label, solution, solution_old, fields, fields_old, bnd_conditions):
+        uv, eta = solution.split()
+        uv_old, eta_old = solution_old.split()
+        return self.residual_cell(label, uv, eta, uv_old, eta_old, fields, fields_old, bnd_conditions)
+
+    def edge_residual(self, label, solution, solution_old, fields, fields_old, bnd_conditions):
+        uv, eta = solution.split()
+        uv_old, eta_old = solution_old.split()
+        return self.residual_edge(label, uv, eta, uv_old, eta_old, fields, fields_old, bnd_conditions)
+
+
+class ShallowWaterMomentumResidual(BaseShallowWaterResidual):
```

Comment:
As this file is getting rather large it might make sense to define the strong residual equation in another file.

Reply:
Thanks, will create a new file for this